ipa-core/src/protocol/ipa_prf/boolean_ops/multiplication.rs

```diff
@@ -1,33 +1,164 @@
-use std::iter::zip;
-
+use crate::protocol::basics::mul::SecureMul;
 use crate::{
     error::Error,
     ff::boolean::Boolean,
     protocol::{
-        context::Context,
-        RecordId,
-        ipa_prf::boolean_ops::{
-            addition_sequential::integer_add,
-        },
+        boolean::NBitStep, context::Context,
+        ipa_prf::boolean_ops::addition_sequential::integer_add, BooleanProtocols, RecordId,
     },
     secret_sharing::{replicated::semi_honest::AdditiveShare, BitDecomposed, FieldSimd},
 };
 
-
+/// This function multiplies x by y in these steps:
+/// 1. Double the input precision and repeat the most significant bit in the extra bits (Sign extension)
+/// 2. Repeatedly multiply x with each digits of y, shift the result 1 digit up each time
+/// 3. Add up the partial products using integer_add
+/// x is assumed to be a positive number
+/// y is assumed to be in two's complement and can be either signed or unsigned
+#[allow(dead_code)]
 pub async fn integer_mul<C, S, const N: usize>(
     ctx: C,
     record_id: RecordId,
     x: &BitDecomposed<AdditiveShare<Boolean, N>>,
     y: &BitDecomposed<AdditiveShare<Boolean, N>>,
-) -> Result<
-    BitDecomposed<AdditiveShare<Boolean, N>>,
-    Error,
->
+) -> Result<BitDecomposed<AdditiveShare<Boolean, N>>, Error>
 where
     C: Context,
     S: NBitStep,
     Boolean: FieldSimd<N>,
     AdditiveShare<Boolean, N>: BooleanProtocols<C, N>,
 {
-    //TODO: To be implemented
+    let new_len = x.len() + y.len();
+    let mut x = x.clone();
+    x.resize(new_len, AdditiveShare::ZERO);
+    let mut y = y.clone();
+    y.resize(new_len, y[y.len() - 1].clone());
+
+    let mut result = BitDecomposed::with_capacity(new_len);
+    for (i, yb) in y.into_iter().enumerate() {
+        let mut t = BitDecomposed::with_capacity(new_len);
+        t.resize(i, AdditiveShare::ZERO);
```

Comment:
Either in this PR (or if you would prefer, in a follow up), we should avoid doing this, as it's needlessly increasing the amount of work in the addition stage, by adding shares of zero. Instead, we should just make `t` of length `new_len - i` in each step, and make sure to add it up correctly.

Reply:
is there a method for left shifting the bits?
i.e.
```
       1101    13
       1011    -5
----------
       1101
     1101      <-- shift left 1 bit
    0000     <-- shift 2 bits
 1101          <-- shift 3 bits
```

right now t.resize is just appending 0s to shift each intermediate products to the left

Reply:
You should be able to create a new BitDecomposed by calling into_iter() on it, chaining that with repeat(ZERO).take(n) in whatever order your shift requires.  That might not be the best option, of course.